### DISCUSSION TITLE
What would it take to render Vega charts without JS?

### DISCUSSION THREAD

**Discussion by huevosabio** [jonmmease/avenger#83] [General]
I've been looking for a good charting library for Dioxus (the Rust UI framework). I feel like between this and Vegafusion it's almost possible to get the full power of the Vega ecosystem without having to ship a JS runtime.

So, how difficult would it be to get to that point?

**Comment by jonmmease** [comment]
Hi @huevosabio, thanks for opening the issue.  This is definitely a goal of mine, and VegaFusion and Avenger are basically the two "ends" of the problem. The next component in the "middle" that I've started working on locally is to implement scales (ala [d3-scales](https://d3js.org/d3-scale)) as Arrow kernels that can be used in DataFusion. Scales are the bridge between the transformed data that VegaFusion generates in data space, that the screen space encodings that Avenger inputs. After scales, another piece that's needed is the logic to convert axis, legend, and colorbar specifications into marks for rendering.

The above will mostly cover what's needed for rendering charts to static images. Introducing interactivity will require integration with a UI framework, like Dioxus, and to implement Vega's [event streams](https://vega.github.io/vega/docs/event-streams/) system for capturing UI events and binding then to Vega signals.

cc @lsh who has been thinking about this as well.

**Comment by huevosabio** [comment]
Gotcha, so sounds like its still quite a bit of work!

I will keep an eye, for now I guess I will have to mix in some JS in any case.

Thanks for this work, I can't wait to be able to ditch JS entirely.

**Reply by jonmmease** [reply]
Are you looking for interactive charts? If not, [vl-convert](https://github.com/vega/vl-convert) is another approach, which converts Vega specs to static images using the regular Vega library, hiding the JavaScript inside Deno.

**Comment by huevosabio** [comment]
Yea, preferably interactive. We're building a companion tool for tracing and logging in games, and I would like it to feel interactive, snappy and portable. The lack of a mature interactive lib in Rust makes me go the Tauri + Svelte route instead of the Dioxus one.

**Comment by airstrike** [comment]
Just wanted to +1 this effort and say that I hope to use this with [`iced`](https://github.com/iced-rs/iced) rather than Dioxus. Part of me thinks it should be fairly easy to write a custom `Widget` for it once the pieces are there.

**Reply by jonmmease** [reply]
Thanks for the +1, it's nice to know where folks are interested in using this outside of the browser.

**Comment by jonmmease** [comment]
out of curiosity, how much do folks care about compatibility with the Vega JSON spec, verses a grammar-based visualization system that's ergonomic to use in Rust directly (maybe something more in the direction of the [Altair API](https://altair-viz.github.io/gallery/index.html) for Rust)?

**Reply by airstrike** [reply]
I don't particularly care about strict compatibility with any Vega JSON stuff tbh. I do think Vega looks quite neat from looking at the website, but I've honestly never used it!

I think the value here is having a way to build interactive\* charts in a declarative way with "batteries included" in Rust. Meaning users won't have to reinvent the wheel to plot bar, line, scatter, {insert your favorite viz}, and pie charts based on some data they have handy.

Currently users can use things like [plotters](https://github.com/plotters-rs/plotters) -- but to my layman eyes that doesn't really feel grammar-based (although it does seem neatly composable)

\* I think the "interactive" part would be the responsibility of the GUI library (`iced`, Dioxus or what have you) and the grammar visualization library just needs to make manipulation accessible. My most basic use cases would be to allow the user to hover over the chart to see more data, drag/scroll the axes, maybe drill down on some data... I don't know, the sky is the limit and that seems far off so I haven't spent a lot of time thinking about this part other than the callouts on hover which to me are table-stakes

**Comment by huevosabio** [comment]
I care first about being able to display interactive charts with Rust without relying on external runtimes. And then I care about whether this is a known JSON standard or just a new ergonomic Rust library.

**Comment by RalfNorthman** [comment]
I just have to say: the thought of having access to vega/vegalite in the Rust ecosystem without any JS sounds wonderful! It would be so nice to have access to a powerful, interactive plotting package in Iced, Bevy, Egui (and Xilem when it has matured).

Don't care about the old json spec, feels much more important with an ergonomic rust crate, preferable one that takes full advantage of the type system to help the viz designer to stay on the right path. EDIT: [Avenger visualization language](https://github.com/avenger-vis/tree-sitter-avenger)? Tell us more!

I have mostly used vegalite via the [gicentre/elm-vegalite](https://dark.elm.dmy.fr/packages/gicentre/elm-vegalite/latest/) package.

**Reply by jonmmease** [reply]
> EDIT: [Avenger visualization language](https://github.com/avenger-vis/tree-sitter-avenger)? Tell us more!

I'm at the very early stages of exploring writing a visualization language/dsl for Avenger that's roughly a combination of [Slint](https://slint.dev/), Vega, and SQL.  The idea is to keep Vega's reactivity and visualization building blocks (scales, axes, etc.), but use a DSL inspired by Slint that makes chart components composable and reusable and provides better IDE support.  And then to use SQL as the expression language rather than Vega's separate expression language and transform system.  This would all be Rust, and the goal would be for it to be embeddable in larger Rust GUI applications.

No promises, but that's where my head is at right now 😄

**Reply by airstrike** [reply]
Since `elm-vegalite` was mentioned, just wanted to flag elm-charts too https://www.elm-charts.org/

And as far as the reusability/composability, it would be nice to keep the DSL optional. I would much rather use "pure Rust" and the builder pattern to construct my charts.

My end goal is embedding charts into an `iced` application and deprecating my [`plotters`](https://github.com/plotters-rs/plotters) dependency
